modules/apps/item-selector/item-selector-taglib/src/main/java/com/liferay/item/selector/taglib/servlet/taglib/ImageSelectorTag.java

```diff
@@ -101,49 +185,200 @@ public void setValidExtensions(String validExtensions) {
 	protected void cleanUp() {
 		super.cleanUp();
 
-		_draggableImage = "none";
 		_fileEntryId = 0;
+		_imageCropDirection = "none";
+		_imageCropRegion = null;
+		_imageURL = null;
 		_itemSelectorEventName = null;
 		_itemSelectorURL = null;
 		_maxFileSize = 0;
+		_namespace = null;
 		_paramName = "imageSelectorFileEntryId";
 		_uploadURL = null;
 		_validExtensions = null;
 	}
 
 	@Override
-	protected String getPage() {
-		return _PAGE;
+	protected String getHydratedModuleName() {
+		return "item-selector-taglib/image_selector/ImageSelector";
 	}
 
 	@Override
-	protected void setAttributes(HttpServletRequest httpServletRequest) {
-		httpServletRequest.setAttribute(
-			"liferay-ui:image-selector:draggableImage", _draggableImage);
-		httpServletRequest.setAttribute(
-			"liferay-ui:image-selector:fileEntryId", _fileEntryId);
-		httpServletRequest.setAttribute(
-			"liferay-ui:image-selector:itemSelectorEventName",
-			_itemSelectorEventName);
-		httpServletRequest.setAttribute(
-			"liferay-ui:image-selector:itemSelectorURL", _itemSelectorURL);
-		httpServletRequest.setAttribute(
-			"liferay-ui:image-selector:maxFileSize", _maxFileSize);
-		httpServletRequest.setAttribute(
-			"liferay-ui:image-selector:paramName", _paramName);
-		httpServletRequest.setAttribute(
-			"liferay-ui:image-selector:uploadURL", _uploadURL);
-		httpServletRequest.setAttribute(
-			"liferay-ui:image-selector:validExtensions", _validExtensions);
-	}
-
-	private static final String _PAGE = "/image_selector/page.jsp";
-
-	private String _draggableImage = "none";
+	protected Map<String, Object> prepareProps(Map<String, Object> props) {
+		props.put("imageCropDirection", _imageCropDirection);
+		props.put("imageCropRegion", _imageCropRegion);
+		props.put("fileEntryId", _fileEntryId);
+		props.put("imageURL", _imageURL);
+		props.put("isDraggable", isDraggable());
+		props.put("itemSelectorEventName", _itemSelectorEventName);
+		props.put("itemSelectorURL", _itemSelectorURL);
+		props.put("maxFileSize", _maxFileSize);
+		props.put("paramName", _paramName);
+		props.put("uploadURL", _uploadURL);
+		props.put("validExtensions", _validExtensions);
+
+		return super.prepareProps(props);
+	}
+
+	@Override
+	protected String processCssClasses(Set<String> cssClasses) {
+		cssClasses.add("drop-zone");
+
+		if (isDraggable()) {
+			cssClasses.add("draggable-image");
+			cssClasses.add(_imageCropDirection);
+		}
+
+		if (_fileEntryId == 0) {
+			cssClasses.add("drop-enabled");
+		}
+
+		cssClasses.add("taglib-image-selector");
+
+		return super.processCssClasses(cssClasses);
+	}
+
+	@Override
+	protected int processStartTag() throws Exception {
```

Comment:
Hi @carloslancha,

As @boton said, I also find the code really hard to read and understand.

I personally think that we should only optimize for performance when we find a performance issue, or when the code is obviously and dangerously slow (which I don't think this would be the case). 

I would measure first, and optimize later. If we can't measure, just use a simpler and easier to read JSP file.

Does FI have plans to avoid this in the future? As we migrate more stuff to React I fear that we will end up with tons of ugly string concatenation like this that JSP already solves.

Reply:
> Does FI have plans to avoid this in the future? As we migrate more stuff to React I fear that we will end up with tons of ugly string concatenation like this that JSP already solves.

Not for now.

I'm fine with what you decide, as I said, I just wanted you to see it done and to explain the pros and cons.

Let me know if you want me to revert the changes and keep the usage of the JSP.

Reply:
I would revert to JSP. But I'll trust whatever @boton and @ambrinchaudhary decide. They are the ones who will maintain it 😝.

Reply:
I do not have a strong opinion.

On one hand, I'm fine with the changes in this taglib because in this way we have one working example with this approach on the other hand I feel more maintainable in JSP.

Personally, I think that avoid this way in apps or taglibs without intensive use as much as possible until we have performance problems or we decide that this is **the way**.

Reply:
I also agree that this way seems more difficult to understand, but I see to good points in favor: and that if we do not know about performance improvements,   but this works as a `SSR` and avoid us to duplicate our html markup just before our <react:component>

Reply:
Up to you guys...

Reply:
I also agree that this way seems more difficult to understand, but I see two good points in favor: 

1) this works as a `SSR` and avoid us to duplicate our html markup just before our `<react:component>`, which also implies more maintenance. 
2) If this really helps about performance, this is always good!

But I really do not understand why some common classes like `NPMResolverProvider`, `ServicesProvider`, `BaseContainerTag` are in this module. Do we have to replicate them in every module? Doesn't make more sense to have it in `frontend-js` or ir some common module, so that the only need for everyone that wants to do something similar would be to write a class extending it (`extends BaseContainerTag`) and avoiding `JSP`? If we move it, I could accept the change, otherwise, I'm afraid I would prefer the old JSPs way ;-). 

@carloslancha @boton @4lejandrito

Reply:
@ambrinchaudhary we talked about that in [this thread](https://liferay.slack.com/archives/CNBG06JS3/p1612860589130400) at #t-frontend-infra

Reply:
Sorry guys, my point 1) is not valid, I didn't see that you are actually writing the "html markup" in the java code, so, yes, this is quite "ugly"... 

Then, I would also revert this code and go in favor of JSP's.... sorry!

Reply:
I view the pros and cons of the two options. So go on with the JSP